torchvision/datasets/inaturalist.py

```diff
@@ -0,0 +1,136 @@
+from PIL import Image
+import os
+import os.path
+from typing import Any, Callable, Dict, List, Optional, Union, Tuple
+
+from .vision import VisionDataset
+from .utils import download_and_extract_archive, verify_str_arg
+
+CATEGORIES = ["kingdom", "phylum", "class", "order", "family", "genus"]
+
+
+class INaturalist(VisionDataset):
+    """`iNaturalist <https://github.com/visipedia/inat_comp>`_ Dataset.
+
+    Args:
+        root (string): Root directory of dataset where the image files are stored.
+            This class does not require/use annotation files.
+        target_type (string or list, optional): Type of target to use, one of:
+            * ``full``: the full category (species)
+            * ``kingdom``: e.g. "Animalia"
+            * ``phylum``: e.g. "Arthropoda"
+            * ``class``: e.g. "Insecta"
+            * ``order``: e.g. "Coleoptera"
+            * ``family``: e.g. "Cleridae"
+            * ``genus``: e.g. "Trichodes"
+            Can also be a list to output a tuple with all specified target types.
+            Defaults to ``full``.
+        transform (callable, optional): A function/transform that takes in an PIL image
+            and returns a transformed version. E.g, ``transforms.RandomCrop``
+        target_transform (callable, optional): A function/transform that takes in the
+            target and transforms it.
+    """
+
+    def __init__(
+            self,
+            root: str,
+            target_type: Union[List[str], str] = "full",
+            transform: Optional[Callable] = None,
+            target_transform: Optional[Callable] = None,
+    ) -> None:
+        super(INaturalist, self).__init__(root,
+                                          transform=transform,
+                                          target_transform=target_transform)
+        if not isinstance(target_type, list):
+            target_type = [target_type]
+        self.target_type = [verify_str_arg(t, "target_type", ("full", *CATEGORIES))
+                            for t in target_type]
+
+        if not os.path.exists(self.root):
+            raise RuntimeError(f'Dataset not found in {self.root}. You need to download it.')
+        self.all_categories = sorted(os.listdir(self.root))
+        if (not self.all_categories):
+            raise RuntimeError(f'Dataset not found in {self.root}. You need to download it.')
```

Comment:
We should provide a `download: bool` flag to handle this. The links can be found [here](https://github.com/visipedia/inat_comp/tree/master/2017#data).

Reply:
The dataset is almost 200GB, I think it might be unfeasible to get it downloaded from the python script running it. Plus, I'm not sure our scripts handle restart after failure.

So in all practicality it might be more feasible to request users to download the dataset manually (otherwise we might end up with a similar situation as ImageNet, where the servers were overloaded by notebooks downloading the dataset each time, and thus blocking use of the data without registration)

Reply:
> The dataset is almost 200GB, I think it might be unfeasible to get it downloaded from the python script running it. 

We already have datasets larger than this (`Places365`, `Kinetics`) so IMO that shouldn't be a reason.

> Plus, I'm not sure our scripts handle restart after failure.

If not we should implement that, albeit not in this PR. At least it should be supported after the rework.

> So in all practicality it might be more feasible to request users to download the dataset manually (otherwise we might end up with a similar situation as ImageNet, where the servers were overloaded by notebooks downloading the dataset each time, and thus blocking use of the data without registration)

Well, ImageNet was different. The download links should have never been public in the first place and I think they just patched that hole. If we simply the verify the archive before trying to download, we shouldn't overloading the server.

Reply:
Ok, let's give this a try then, although I think we will need more robust download functionality for such large-scale datasets

Reply:
One thing to note is that there are several datasets, 2017, 2018, etc. I'll make it so the user specifies one, instead of a boolean argument.